starknet/src/library/mmr_verifier.cairo

```diff
@@ -0,0 +1,801 @@
+            }
+            elements_count -= mountain_elements_count;
+            element_index -= mountain_elements_count;
+            mountain_index += 1;
+        }
+        mountain_elements_count = BitShift::shr(mountain_elements_count, 1);
+        mountain_height -= 1;
+    };
+    result
+}
+
+fn element_index_to_leaf_index(element_index: usize) -> Result<usize, felt252> {
+    if element_index == 0 {
+        panic!("Invalid element index");
+    }
+    elements_count_to_leaf_count(element_index - 1)
+}
+
+fn elements_count_to_leaf_count(elements_count: usize) -> Result<usize, felt252> {
+    let mut leaf_count = 0;
+    let mut mountain_leaf_count = BitShift::shl(1, bit_length(elements_count));
+    let mut current_elements_count = elements_count;
+
+    while mountain_leaf_count > 0 {
+        let mountain_elements_count = 2 * mountain_leaf_count - 1;
+        if mountain_elements_count <= current_elements_count {
+            leaf_count += mountain_leaf_count;
+            current_elements_count -= mountain_elements_count;
+        }
+        mountain_leaf_count = BitShift::shr(mountain_leaf_count, 1);
+    };
+
+    if current_elements_count > 0 {
+        return Result::Err('Invalid elements count');
+    } else {
+        return Result::Ok(leaf_count);
+    }
+}
+
+fn retrieve_peaks_hashes(peak_idxs: Span<u256>) -> Result<Array<u256>, felt252> {
+    Result::Ok(array![])
+}
+
+fn find_peaks(mut elements_count: usize) -> Array<usize> {
+    let mut mountain_elements_count = (BitShift::shl(1, bit_length(elements_count))) - 1;
+    let mut mountain_index_shift = 0;
+    let mut peaks: Array<usize> = array![];
+
+    while mountain_elements_count > 0 {
+        if mountain_elements_count <= elements_count {
+            mountain_index_shift += mountain_elements_count;
+            peaks.append(mountain_index_shift);
+            elements_count -= mountain_elements_count;
+        }
+        mountain_elements_count = BitShift::shr(mountain_elements_count, 1);
+    };
+
+    if elements_count > 0 {
+        return array![];
+    }
+
+    peaks
+}
+
+
+fn bit_length(num: usize) -> usize {
+    if num == 0 {
+        return 0;
+    }
+
+    let mut bit_position = 0;
+    let mut curr_n = 1;
+    while num >= curr_n {
+        bit_position += 1;
+        curr_n = BitShift::shl(curr_n, 1);
+    };
+    bit_position
+}
+
+pub fn extract_state_root(rlp_encoded: Span<u8>) -> u256 {
+    let start_index = 90;
+    let length = 32;
+    let mut state_root_bytes = array![];
+
+    let mut i = start_index;
+    while i <= start_index + length {
+        state_root_bytes.append(*rlp_encoded.at(i));
+        i += 1;
+    };
+
+    let len = state_root_bytes.len();
+    let mut j = 0;
+    let mut state_root: u256 = 0;
+
+    while j < len {
+        state_root = (BitShift::shl(state_root, 8)) + (*state_root_bytes.at(j)).into();
+        j += 1;
+    };
+
+    state_root
+}
+
+pub fn extract_block_number(rlp_encoded: Span<u8>) -> u64 {
```

Comment:
Remove this, we can't extract the block numbers from the rlp because their length is variable, block number will be sent as parameter